<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Convenient wrapper functions for the analysis of
    matrix-assisted laser desorption/ionization-time-of-flight (MALDI-TOF)
    spectra data in order to select only representative spectra (also
    called cherry-pick). The package covers the preprocessing and
    dereplication steps (based on Strejcek, Smrhova, Junkova and Uhlik
    (2018) &lt;doi:10.3389/fmicb.2018.01294&gt;) needed to cluster MALDI-TOF
    spectra before the final cherry-picking step. It enables the easy
    exclusion of spectra and/or clusters to accommodate complex
    cherry-picking strategies. Alternatively, cherry-picking using
    taxonomic identification MALDI-TOF data is made easy with functions to
    import inconsistently formatted reports.">
<title>Dereplicate and Cherry-Pick Mass Spectrometry Spectra • maldipickr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Dereplicate and Cherry-Pick Mass Spectrometry Spectra">
<meta property="og:description" content="Convenient wrapper functions for the analysis of
    matrix-assisted laser desorption/ionization-time-of-flight (MALDI-TOF)
    spectra data in order to select only representative spectra (also
    called cherry-pick). The package covers the preprocessing and
    dereplication steps (based on Strejcek, Smrhova, Junkova and Uhlik
    (2018) &lt;doi:10.3389/fmicb.2018.01294&gt;) needed to cluster MALDI-TOF
    spectra before the final cherry-picking step. It enables the easy
    exclusion of spectra and/or clusters to accommodate complex
    cherry-picking strategies. Alternatively, cherry-picking using
    taxonomic identification MALDI-TOF data is made easy with functions to
    import inconsistently formatted reports.">
<meta property="og:image" content="https://clavellab.github.io/maldipickr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">maldipickr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/dereplicate-bruker-maldi-biotyper-spectra.html">Dereplicate Bruker MALDI Biotyper spectra</a>
    <a class="dropdown-item" href="articles/import-data-from-bruker-maldi-biotyper.html">Import data from Bruker MALDI Biotyper</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <h6 class="dropdown-header" data-toc-skip>Releases</h6>
    <a class="dropdown-item" href="https://clavellab.github.io/maldipickr/news/index.html#maldipickr-120">Version 1.2.0</a>
    <a class="dropdown-item" href="https://clavellab.github.io/maldipickr/news/index.html#maldipickr-111">Version 1.1.1</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="news/index.html">Changelog</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ClavelLab/maldipickr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="maldipickr-">maldipickr <a class="anchor" aria-label="anchor" href="#maldipickr-"></a>
</h1>
</div>
<!-- badges: start -->

<ul>
<li>Are you using the MALDI-TOF<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Matrix-Assisted Laser Desorption/Ionization-Time-Of-Flight (MALDI-TOF)&lt;/p&gt;"><sup>1</sup></a> Biotyper to identify bacterial isolates? <strong>Yes</strong>
</li>
<li>Do you want to select representative isolates for further experiments? <strong>Yes</strong>
</li>
<li>Do you need fast and automated selection decisions that you can retrace? <strong>Yes</strong>
</li>
</ul>
<p>↪ <em>The <a href="https://github.com/ClavelLab/maldipickr" class="external-link"><code>{maldipickr}</code></a> package is right for your needs!</em> The documented and tested R functions will help you dereplicate MALDI-TOF data and cherry-pick representative spectra of microbial isolates.</p>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<p>How to <strong>cherry-pick bacterial isolates</strong> with MALDI Biotyper:</p>
<ul>
<li><a href="#using-spectra-data">using spectra data</a></li>
<li><a href="#using-taxonomic-identification-report">using taxonomic identification report</a></li>
</ul>
<div class="section level3">
<h3 id="using-spectra-data">Using spectra data<a class="anchor" aria-label="anchor" href="#using-spectra-data"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ClavelLab/maldipickr" class="external-link">maldipickr</a></span><span class="op">)</span></span>
<span><span class="co"># Set up the directory location of your spectra data</span></span>
<span><span class="va">spectra_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"toy-species-spectra"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import and process the spectra</span></span>
<span><span class="va">processed</span> <span class="op">&lt;-</span> <span class="va">spectra_dir</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/import_biotyper_spectra.html">import_biotyper_spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/process_spectra.html">process_spectra</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Delineate spectra clusters using Cosine similarity</span></span>
<span><span class="co">#  and cherry-pick one representative spectra.</span></span>
<span><span class="co">#  The chosen ones are indicated by `to_pick` column</span></span>
<span><span class="va">processed</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/merge_processed_spectra.html">merge_processed_spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">coop</span><span class="fu">::</span><span class="fu">tcosine</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/delineate_with_similarity.html">delineate_with_similarity</a></span><span class="op">(</span>threshold <span class="op">=</span> <span class="fl">0.92</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/set_reference_spectra.html">set_reference_spectra</a></span><span class="op">(</span><span class="va">processed</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">to_pick</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 7</span></span>
<span><span class="co">#&gt;   name         to_pick membership cluster_size   SNR peaks is_reference</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;        &lt;lgl&gt;        &lt;int&gt;        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;       </span></span>
<span><span class="co">#&gt; 1 species1_G2  FALSE            1            4  5.09    21 FALSE       </span></span>
<span><span class="co">#&gt; 2 species2_E11 FALSE            2            2  5.54    22 FALSE       </span></span>
<span><span class="co">#&gt; 3 species2_E12 TRUE             2            2  5.63    23 TRUE        </span></span>
<span><span class="co">#&gt; 4 species3_F7  FALSE            1            4  4.89    26 FALSE       </span></span>
<span><span class="co">#&gt; 5 species3_F8  TRUE             1            4  5.56    25 TRUE        </span></span>
<span><span class="co">#&gt; 6 species3_F9  FALSE            1            4  5.40    25 FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-taxonomic-identification-report">Using taxonomic identification report<a class="anchor" aria-label="anchor" href="#using-taxonomic-identification-report"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ClavelLab/maldipickr" class="external-link">maldipickr</a></span><span class="op">)</span></span>
<span><span class="co"># Import Biotyper CSV report</span></span>
<span><span class="co">#  and glimpse at the table</span></span>
<span><span class="va">report_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_biotyper_report.html">read_biotyper_report</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"biotyper_unknown.csv"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">report_tbl</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">bruker_species</span>, <span class="va">bruker_log</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 4 × 3</span></span>
<span><span class="co">#&gt;   name              bruker_species               bruker_log</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;                             &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 unknown_isolate_1 not reliable identification        1.33</span></span>
<span><span class="co">#&gt; 2 unknown_isolate_2 not reliable identification        1.4 </span></span>
<span><span class="co">#&gt; 3 unknown_isolate_3 Faecalibacterium prausnitzii       1.96</span></span>
<span><span class="co">#&gt; 4 unknown_isolate_4 Faecalibacterium prausnitzii       2.07</span></span>
<span></span>
<span><span class="co"># Delineate clusters from the identifications</span></span>
<span><span class="co">#   and cherry-pick one representative spectra.</span></span>
<span><span class="co">#   The chosen ones are indicated by `to_pick` column</span></span>
<span><span class="va">report_tbl</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/delineate_with_identification.html">delineate_with_identification</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="va">report_tbl</span>, criteria_column <span class="op">=</span> <span class="st">"bruker_log"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">to_pick</span>, <span class="va">bruker_species</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating clusters from single report</span></span>
<span><span class="co">#&gt; # A tibble: 4 × 11</span></span>
<span><span class="co">#&gt;   name       to_pick bruker_species membership cluster_size sample_name hit_rank</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;      &lt;lgl&gt;   &lt;chr&gt;               &lt;int&gt;        &lt;int&gt; &lt;chr&gt;          &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 unknown_i… TRUE    not reliable …          2            1 &lt;NA&gt;               1</span></span>
<span><span class="co">#&gt; 2 unknown_i… TRUE    not reliable …          3            1 &lt;NA&gt;               1</span></span>
<span><span class="co">#&gt; 3 unknown_i… FALSE   Faecalibacter…          1            2 &lt;NA&gt;               1</span></span>
<span><span class="co">#&gt; 4 unknown_i… TRUE    Faecalibacter…          1            2 &lt;NA&gt;               1</span></span>
<span><span class="co">#&gt; # ℹ 4 more variables: bruker_quality &lt;chr&gt;, bruker_taxid &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   bruker_hash &lt;chr&gt;, bruker_log &lt;dbl&gt;</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><a href="https://github.com/ClavelLab/maldipickr" class="external-link">maldipickr</a> is available on the <a href="https://cran.r-project.org/package=maldipickr" class="external-link">CRAN</a> and on <a href="https://github.com/ClavelLab/maldipickr" class="external-link">GitHub</a>.</p>
<p>To install the latest CRAN release, use the following command in R:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"maldipickr"</span><span class="op">)</span></span></code></pre></div>
<p>To install the development version, use the following command in R:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ClavelLab/maldipickr"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>The comprehensive vignettes will walk you through the package functions and showcase how to:</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://clavellab.github.io/maldipickr/articles/import-data-from-bruker-maldi-biotyper.html">Import spectra data and identification reports from Bruker MALDI Biotyper into R</a>.</li>
<li>
<a href="https://clavellab.github.io/maldipickr/articles/dereplicate-bruker-maldi-biotyper-spectra.html">Process, dereplicate and cherry-pick representative spectra, from simple to complex design</a>.</li>
</ol>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>This R package is developed for spectra data generated by the Bruker MALDI Biotyper device. The <a href="https://github.com/ClavelLab/maldipickr" class="external-link"><code>{maldipickr}</code></a> package is built from a suite of Rmarkdown files using the <a href="https://thinkr-open.github.io/fusen/" class="external-link"><code>{fusen}</code></a> package by Rochette S (2023). It relies on:</p>
<ol style="list-style-type: decimal">
<li>the <a href="https://cran.r-project.org/package=MALDIquant" class="external-link"><code>{MALDIquant}</code></a> package from Gibb &amp; Strimmer (2012) for spectra functions</li>
<li>the work of Strejcek et al. (2018) for the dereplication procedure.</li>
</ol>
<div class="section level3">
<h3 id="disclaimer">Disclaimer<a class="anchor" aria-label="anchor" href="#disclaimer"></a>
</h3>
<p>The developers of this package are part of the <a href="https://www.ukaachen.de/kliniken-institute/institut-fuer-medizinische-mikrobiologie/forschung/ag-clavel/" class="external-link">Clavel Lab</a> and are not affiliated with the company Bruker, therefore this package is independent of the company and is distributed under the <a href="https://clavellab.github.io/maldipickr/LICENSE.html">GPL-3.0 License</a>.</p>
<p>The hexagonal logo was created by Charlie Pauvert and uses the <a href="https://sourcefoundry.org/hack/" class="external-link">Hack</a> font and a color palette generated at <a href="https://coolors.co" class="external-link uri">https://coolors.co</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Gibb S &amp; Strimmer K (2012). “MALDIquant: a versatile R package for the analysis of mass spectrometry data”. <em>Bioinformatics</em> 28, 2270-2271. <a href="https://doi.org/10.1093/bioinformatics/bts447" class="external-link uri">https://doi.org/10.1093/bioinformatics/bts447</a>.</li>
<li>Rochette S (2023). “fusen: Build a Package from Rmarkdown Files”. <a href="https://thinkr-open.github.io/fusen/" class="external-link uri">https://thinkr-open.github.io/fusen/</a>, <a href="https://github.com/Thinkr-open/fusen" class="external-link uri">https://github.com/Thinkr-open/fusen</a>.</li>
<li>Strejcek M, Smrhova T, Junkova P &amp; Uhlik O (2018). “Whole-Cell MALDI-TOF MS versus 16S rRNA Gene Analysis for Identification and Dereplication of Recurrent Bacterial Isolates.” <em>Frontiers in Microbiology</em> 9 <a href="https://doi.org/10.3389/fmicb.2018.01294" class="external-link uri">https://doi.org/10.3389/fmicb.2018.01294</a>.</li>
</ul>
</div>
</div>


  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=maldipickr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/ClavelLab/maldipickr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ClavelLab/maldipickr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing maldipickr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Charlie Pauvert <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0001-9832-2507" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Thomas Clavel <br><small class="roles"> Contributor, funder, copyright holder </small> <a href="https://orcid.org/0000-0002-7229-5595" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://CRAN.R-project.org/package=maldipickr" class="external-link"><img src="https://www.r-pkg.org/badges/version/maldipickr" alt="CRAN status"></a></li>
<li><a href="https://www.repostatus.org/#active" class="external-link"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://github.com/ClavelLab/maldipickr/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/ClavelLab/maldipickr/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/github/ClavelLab/maldipickr" class="external-link"><img src="https://codecov.io/github/ClavelLab/maldipickr/branch/main/graph/badge.svg?token=JQABKDK2MB" alt="codecov"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Charlie Pauvert, Thomas Clavel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
