<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="maldipickr">
<title>Dereplicate Bruker MALDI Biotyper spectra • maldipickr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Dereplicate Bruker MALDI Biotyper spectra">
<meta property="og:description" content="maldipickr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">maldipickr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a-import-data-from-bruker-maldi-biotyper.html">Import data from Bruker MALDI Biotyper</a>
    <a class="dropdown-item" href="../articles/b-dereplicate-bruker-maldi-biotyper-spectra.html">Dereplicate Bruker MALDI Biotyper spectra</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cpauvert/maldipickr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Dereplicate Bruker MALDI Biotyper spectra</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cpauvert/maldipickr/blob/HEAD/vignettes/b-dereplicate-bruker-maldi-biotyper-spectra.Rmd" class="external-link"><code>vignettes/b-dereplicate-bruker-maldi-biotyper-spectra.Rmd</code></a></small>
      <div class="d-none name"><code>b-dereplicate-bruker-maldi-biotyper-spectra.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cpauvert/maldipickr" class="external-link">maldipickr</a></span><span class="op">)</span></span></code></pre></div>
<!-- WARNING - This vignette is generated by {fusen} from dev/dereplicate-spectra.Rmd: do not edit by hand -->
<p>Bacterial colony identification with the Bruker MALDI Biotyper is a
high-throughput method with the built-in tools, provided that the
selected bacteria belong to the internal database.</p>
<p>Scientific projects where the number of unknown bacteria is expected
to be high needs reference-free methods to be able to reduce the
redundancy of isolated bacterial colonies, a process called
<em>dereplication</em>.</p>
<p><a href="https://doi.org/10.3389/fmicb.2018.01294" class="external-link">Strejcek <em>et
al.</em> (2018)</a> proposed such a method by processing the spectra and
suggest similarity thresholds between spectra above which spectra, and
therefore the measured bacterial colonies, can be considered identical
at a given taxonomic rank. Their processing procedure is implemented in
the <a href="https://github.com/cpauvert/maldipickr" class="external-link"><code>maldipickr</code></a>
package and illustrated in the following vignette.</p>
<p>In addition, we provide functions to enable the dereplication of
different batches of Bruker MALDI Biotyper runs and combine the results,
in order to be able to delineate the clusters from a common similarity
matrix.</p>
<p>More importantly, we provide a function to select a spectra to be
picked in each cluster, a process called <em>cherry-picking</em>,
depending on external metadata and potential out-groups to be excluded
for the current cherry-picking steps.</p>
<div class="section level2">
<h2 id="process-bruker-maldi-biotyper-spectra">Process Bruker MALDI Biotyper spectra<a class="anchor" aria-label="anchor" href="#process-bruker-maldi-biotyper-spectra"></a>
</h2>
<p>From the imported raw data from the Bruker MALDI Biotyper, the
processing of the spectra is based on the original implementation, and
run the following tasks:</p>
<ol style="list-style-type: decimal">
<li>Square-root transformation</li>
<li>Mass range trimming to 4-10 kDa as they were deemed most determinant
by Strejcek et al. (2018)</li>
<li>Signal smoothing using the Savitzky-Golay method and a half window
size of 20</li>
<li>Baseline correction with the SNIP procedure</li>
<li>Normalisation by Total Ion Current</li>
<li>Peak detection using the SuperSmoother procedure and with a
signal-to-noise ratio above 3</li>
<li>Peak filtering. This step has been added to discard peaks with a
negative signal-to-noise ratio probably due to being on the edge of the
mass range.</li>
</ol>
<p>The full procedure is illustrated in the example below. While in this
case, all the resulting processed spectra, peaks and final spectra
metadata are stored in-memory, the <code>process_spectra</code> function
enables storing these files locally for scalable high-throughput
analyses.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span><span class="va">directory_for_biotyper_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"toy-species-spectra"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span></span>
<span><span class="co"># Import the six spectra</span></span>
<span><span class="va">spectra_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_biotyper_spectra.html">import_biotyper_spectra</a></span><span class="op">(</span><span class="va">directory_for_biotyper_spectra</span><span class="op">)</span></span>
<span><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span><span class="va">processed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_spectra.html">process_spectra</a></span><span class="op">(</span><span class="va">spectra_list</span><span class="op">)</span></span>
<span><span class="co"># Overview of the list architecture that is returned</span></span>
<span><span class="co">#  with the list of processed spectra, peaks identified and the</span></span>
<span><span class="co">#  metadata table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">processed</span>, max.level <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ spectra :List of 6</span></span>
<span><span class="co">#&gt;   ..$ species1_G2 :Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;   ..$ species2_E11:Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;   ..$ species2_E12:Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F7 :Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F8 :Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F9 :Formal class 'MassSpectrum' [package "MALDIquant"] with 3 slots</span></span>
<span><span class="co">#&gt;  $ peaks   :List of 6</span></span>
<span><span class="co">#&gt;   ..$ species1_G2 :Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..$ species2_E11:Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..$ species2_E12:Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F7 :Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F8 :Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..$ species3_F9 :Formal class 'MassPeaks' [package "MALDIquant"] with 4 slots</span></span>
<span><span class="co">#&gt;  $ metadata: tibble [6 × 3] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co"># A detailed view of the metadata with the median signal-to-noise</span></span>
<span><span class="co">#  ratio (SNR) and the number of peaks</span></span>
<span><span class="va">processed</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   name           SNR peaks</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2   5.09    21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11  5.54    22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12  5.63    23</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7   4.89    26</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8   5.56    25</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9   5.40    25</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="merge-multiples-processed-spectra">Merge multiples processed spectra<a class="anchor" aria-label="anchor" href="#merge-multiples-processed-spectra"></a>
</h2>
<p>During high-throughput analyses, multiples runs of Bruker MALDI
Biotyper are expected resulting in several batches of spectra to be
processed and compared. While their processing is natively independent,
and could natively be run in parallell, the integration of the batches
for their comparison needs an additional step.</p>
<p>The <code>merge_processed_spectra</code> function aggregates the
processed spectra and bins together the detected peaks, with a tolerance
of 2e-09 ppm. This binning step results in a <span class="math inline">\(n\times p\)</span> feature matrix (or intensity
matrix), with <span class="math inline">\(n\)</span> rows for <span class="math inline">\(n\)</span> processed spectra (peakless spectra are
discarded) and <span class="math inline">\(p\)</span> columns for the
<span class="math inline">\(p\)</span> peaks masses.</p>
<p>By default, as in the Strejeck et al. (2018) procedure, the intensity
values for spectra with missing peaks are interpolated from the
processed spectra signal. The current function enables the analyst to
decide whether to interpolate the values or leave missing peaks as
<code>NA</code> which would then be converted to an null intensity
value.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span><span class="va">directory_for_biotyper_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"toy-species-spectra"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span></span>
<span><span class="co"># Import the six spectra</span></span>
<span><span class="va">spectra_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_biotyper_spectra.html">import_biotyper_spectra</a></span><span class="op">(</span><span class="va">directory_for_biotyper_spectra</span><span class="op">)</span></span>
<span><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span><span class="va">processed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_spectra.html">process_spectra</a></span><span class="op">(</span><span class="va">spectra_list</span><span class="op">)</span></span>
<span><span class="co"># Merge the spectra to produce the feature matrix</span></span>
<span><span class="va">fm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_processed_spectra.html">merge_processed_spectra</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">processed</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># The feature matrix has 6 spectra as rows and</span></span>
<span><span class="co">#  35 peaks as columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  6 35</span></span>
<span><span class="co"># Notice the difference when the interpolation is turned off</span></span>
<span><span class="va">fm_no_interpolation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_processed_spectra.html">merge_processed_spectra</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">processed</span><span class="op">)</span>, interpolate_missing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">fm</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># 0</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">fm_no_interpolation</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># 68</span></span>
<span><span class="co">#&gt; [1] 68</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compute-a-similarity-matrix-between-all-processed-spectra-not-included">Compute a similarity matrix between all processed spectra (not
included)<a class="anchor" aria-label="anchor" href="#compute-a-similarity-matrix-between-all-processed-spectra-not-included"></a>
</h2>
<p>Once all the batches of spectra have been processed together, we can
use a distance metric to evaluate how close the spectra are to one
another. <a href="https://doi.org/10.3389/fmicb.2018.01294" class="external-link">Strejcek
<em>et al.</em> (2018)</a> recommend the <em>cosine</em> metric to
compare the spectra and they use the fast implementation in the
<code>coop</code> package.</p>
<p>While we do not provide specific functions to generate the similarity
matrix, we illustrate below how it can be easily computed. Note that the
feature matrix from <code>merge_processed_spectra</code> has spectra as
rows and peaks values as columns. So to get a similarity matrix between
spectra, the feature matrix must be transposed before cosine
computation.</p>
<pre><code><span></span>
<span><span class="co"># Install the coop package</span></span>
<span></span>
<span><span class="co"># install.packages("coop")</span></span>
<span></span>
<span><span class="co"># Compute the similarity matrix on the transposed feature matrix</span></span>
<span></span>
<span><span class="va">sim_matrix</span> <span class="op">&lt;-</span> <span class="fu">coop</span><span class="fu">::</span><span class="fu">cosine</span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="delineate-clusters-from-a-similarity-matrix">Delineate clusters from a similarity matrix<a class="anchor" aria-label="anchor" href="#delineate-clusters-from-a-similarity-matrix"></a>
</h2>
<p>When the similarity matrix is computed between all pairs of the
studied spectra, the next step is to delineate clusters of spectra to
dereplicate the measured bacterial colonies.</p>
<p>The <code>similarity_to_clusters</code> is agnostic of the similarity
metric used, whether it is the cosine metric or the Pearson product
moment, provided that a numeric threshold relevant to the metric used is
given.</p>
<p>Indeed, the matrix is transformed into a network without loops, where
nodes are spectra and links exist between spectra only if the similarity
between the spectra is <strong>above</strong> (or <strong>equal
to</strong>) the threshold. This representation allows to infer the
clusters. A table summarises for each spectra, to which cluster number
it was assigned to and the size of the cluster, which is the total
number of spectra in the cluster.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Toy similarity matrix between the six example spectra of</span></span>
<span><span class="co">#  three species. The cosine metric is used and a value of</span></span>
<span><span class="co">#  zero indicates dissimilar spectra and a value of one</span></span>
<span><span class="co">#  indicates identical spectra.</span></span>
<span><span class="va">cosine_similarity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fl">1</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">0.98</span>, <span class="fl">0.98</span>,</span>
<span>    <span class="fl">0.79</span>, <span class="fl">1</span>, <span class="fl">0.98</span>, <span class="fl">0.79</span>, <span class="fl">0.8</span>, <span class="fl">0.8</span>,</span>
<span>    <span class="fl">0.77</span>, <span class="fl">0.98</span>, <span class="fl">1</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>,</span>
<span>    <span class="fl">0.99</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.99</span>,</span>
<span>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">1</span>, <span class="fl">1</span></span>
<span>  <span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"species1_G2"</span>, <span class="st">"species2_E11"</span>, <span class="st">"species2_E12"</span>,</span>
<span>      <span class="st">"species3_F7"</span>, <span class="st">"species3_F8"</span>, <span class="st">"species3_F9"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"species1_G2"</span>, <span class="st">"species2_E11"</span>, <span class="st">"species2_E12"</span>,</span>
<span>      <span class="st">"species3_F7"</span>, <span class="st">"species3_F8"</span>, <span class="st">"species3_F9"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Delineate clusters based on a 0.92 threshold applied</span></span>
<span><span class="co">#  to the similarity matrix</span></span>
<span><span class="fu"><a href="../reference/similarity_to_clusters.html">similarity_to_clusters</a></span><span class="op">(</span><span class="va">cosine_similarity</span>, threshold <span class="op">=</span> <span class="fl">0.92</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11          2            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12          2            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           1            4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           1            4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           1            4</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-a-reference-spectrum-for-each-cluster">Set a reference spectrum for each cluster<a class="anchor" aria-label="anchor" href="#set-a-reference-spectrum-for-each-cluster"></a>
</h2>
<p>Once the table of clusters is generated from the similarity matrix, a
reference spectrum can be assigned to each cluster.</p>
<p>We choose to define high-quality spectra as representative spectra of
the clusters using internal information. That is, representative spectra
have, within their cluster, the highest median signal-to-noise ratio and
then the highest number of detected peaks.</p>
<p>The function <code>set_reference_spectra</code> does not change the
order of the cluster table but merely add an additional column
<code>is_reference</code> to indicate if the corresponding spectrum is
representative of the cluster.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get an example directory of six Bruker MALDI Biotyper spectra</span></span>
<span><span class="co"># Import the six spectra and</span></span>
<span><span class="co"># Transform the spectra signals according to Strejcek et al. (2018)</span></span>
<span><span class="va">processed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"toy-species-spectra"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/import_biotyper_spectra.html">import_biotyper_spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/process_spectra.html">process_spectra</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Toy similarity matrix between the six example spectra of</span></span>
<span><span class="co">#  three species. The cosine metric is used and a value of</span></span>
<span><span class="co">#  zero indicates dissimilar spectra and a value of one</span></span>
<span><span class="co">#  indicates identical spectra.</span></span>
<span><span class="va">cosine_similarity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fl">1</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">0.98</span>, <span class="fl">0.98</span>,</span>
<span>    <span class="fl">0.79</span>, <span class="fl">1</span>, <span class="fl">0.98</span>, <span class="fl">0.79</span>, <span class="fl">0.8</span>, <span class="fl">0.8</span>,</span>
<span>    <span class="fl">0.77</span>, <span class="fl">0.98</span>, <span class="fl">1</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>, <span class="fl">0.77</span>,</span>
<span>    <span class="fl">0.99</span>, <span class="fl">0.79</span>, <span class="fl">0.77</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.99</span>,</span>
<span>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>    <span class="fl">0.98</span>, <span class="fl">0.8</span>, <span class="fl">0.77</span>, <span class="fl">0.99</span>, <span class="fl">1</span>, <span class="fl">1</span></span>
<span>  <span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"species1_G2"</span>, <span class="st">"species2_E11"</span>, <span class="st">"species2_E12"</span>,</span>
<span>      <span class="st">"species3_F7"</span>, <span class="st">"species3_F8"</span>, <span class="st">"species3_F9"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"species1_G2"</span>, <span class="st">"species2_E11"</span>, <span class="st">"species2_E12"</span>,</span>
<span>      <span class="st">"species3_F7"</span>, <span class="st">"species3_F8"</span>, <span class="st">"species3_F9"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Delineate clusters based on a 0.92 threshold applied</span></span>
<span><span class="co">#  to the similarity matrix</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/similarity_to_clusters.html">similarity_to_clusters</a></span><span class="op">(</span><span class="va">cosine_similarity</span>, threshold <span class="op">=</span> <span class="fl">0.92</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set reference spectra with the toy example</span></span>
<span><span class="fu"><a href="../reference/set_reference_spectra.html">set_reference_spectra</a></span><span class="op">(</span><span class="va">clusters</span>, <span class="va">processed</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size   SNR peaks is_reference</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            4  5.09    21 FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11          2            2  5.54    22 FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12          2            2  5.63    23 TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           1            4  4.89    26 FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           1            4  5.56    25 TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           1            4  5.40    25 FALSE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="import-clusters-results-generated-by-spede">Import clusters results generated by SPeDE<a class="anchor" aria-label="anchor" href="#import-clusters-results-generated-by-spede"></a>
</h2>
<p>Raw spectra can also be processed and clustered by another approach,
named <a href="https://doi.org/10.1128/mSystems.00437-19" class="external-link"><code>SPeDE</code></a>,
developed by Dumolin et al. (2019). The resulting dereplication step
produces a comma separated table. The example below illustrates how to
import this table into R to be consistent with the dereplication table
generated within the <a href="https://github.com/cpauvert/maldipickr" class="external-link"><code>maldipickr</code></a>
package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reformat the output from SPeDE table</span></span>
<span><span class="co"># https://github.com/LM-UGent/SPeDE</span></span>
<span><span class="fu"><a href="../reference/import_spede_clusters.html">import_spede_clusters</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"spede.csv"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size quality is_reference</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            1 GREEN   TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11          2            2 ORANGE  FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12          2            2 GREEN   TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           3            1 GREEN   TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           4            2 ORANGE  FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           4            2 GREEN   TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cherry-pick-bruker-maldi-biotyper-spectra">Cherry-pick Bruker MALDI Biotyper spectra<a class="anchor" aria-label="anchor" href="#cherry-pick-bruker-maldi-biotyper-spectra"></a>
</h2>
<p>When isolating bacteria from an environment, experimenters want to be
thorough but also work-, time- and cost-savvy. One approach is to reduce
the redundancy of the bacterial isolates by analyzing their MALDI-TOF
spectra from the Bruker Biotyper. All the steps previously described in
this vignette consisted of processing the spectra to be able to pick
only non-redundant spectra, using the <code>pick_spectra</code>
function.</p>
<p>The function, as illustrated in the example below, can pick spectra
using different types of inputs:</p>
<ul>
<li>the reference spectra information that is present in the cluster
table (after using <code>similarity_to_clusters</code> or
<code>import_spede_clusters</code> functions)</li>
<li>an external metadata table containing a variable (e.g., optical
density, fluorescence) to be maximized (default) or minimized per
cluster</li>
<li>a regular expression applied to the spectra names to explicitly
removed some spectra and therefore clusters from the picking decision.
This is particularly useful if some spectra have been previously picked.
For example, the regular expression could match a date or time to
exclude colonies grown and picked 24h after streaking when comparing
with colonies grown for 72h, or match a growth condition such as a
temperature or an antibiotic provided this information is in the spectra
name. A dry-run option is provided to display the spectra to be removed
with the regular expression <em>without actually performing the picking
decision</em>.</li>
</ul>
<p>Advanced users can also provide directly a cluster table with a
custom sort by cluster to accommodate complex design.</p>
<p>Ultimately, the function delivers a table with as many rows as the
cluster table with an additional logical column named
<code>to_pick</code> to indicate whether the colony associated with the
spectra should be picked (<code>TRUE</code>) or not picked
(<code>FALSE</code>).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load a toy example of a tibble of clusters created by the `similarity_to_clusters`</span></span>
<span><span class="co">#  function.</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"clusters_tibble.RDS"</span>, package <span class="op">=</span> <span class="st">"maldipickr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># By default and if no other metadata are provided, the function</span></span>
<span><span class="co">#  picks the reference spectrum of each clusters.</span></span>
<span><span class="fu"><a href="../reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size   SNR peaks is_reference to_pick</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            4  5.09    21 FALSE        FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11          2            2  5.54    22 FALSE        FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12          2            2  5.63    23 TRUE         TRUE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           1            4  4.89    26 FALSE        FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           1            4  5.56    25 TRUE         TRUE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           1            4  5.40    25 FALSE        FALSE</span></span>
<span></span>
<span><span class="co"># Simulate OD600 values with uniform distribution</span></span>
<span><span class="co">#  for each of the colonies we measured with</span></span>
<span><span class="co">#  the Bruker MALDI Biotyper</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">104</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/transmute.html" class="external-link">transmute</a></span><span class="op">(</span><span class="va">clusters</span>, name <span class="op">=</span> <span class="va">name</span>, OD600 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">metadata</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   name         OD600</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2  0.364</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11 0.772</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12 0.735</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7  0.973</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8  0.740</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9  0.201</span></span>
<span></span>
<span><span class="co"># Pick the spectra based on the highest OD600 value per cluster</span></span>
<span><span class="fu"><a href="../reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="va">clusters</span>, <span class="va">metadata</span>, <span class="st">"OD600"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size   SNR peaks is_reference OD600 to_pick</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            4  5.09    21 FALSE        0.364 FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11          2            2  5.54    22 FALSE        0.772 TRUE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12          2            2  5.63    23 TRUE         0.735 FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           1            4  4.89    26 FALSE        0.973 TRUE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           1            4  5.56    25 TRUE         0.740 FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           1            4  5.40    25 FALSE        0.201 FALSE</span></span>
<span></span>
<span><span class="co"># Say that spectra 'species2_E11' was picked before in a previous experiment</span></span>
<span><span class="co">#  we do not want to pick a cluster where species2_E11 is included to limit redundancy</span></span>
<span><span class="co"># The function can peform a dry run to indicate which spectra would be removed</span></span>
<span><span class="fu"><a href="../reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="va">clusters</span>, discard_regex <span class="op">=</span> <span class="st">"E11"</span>, only_show_discarded <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Returning more (or less) than 1 row per `summarise()` group was deprecated in</span></span>
<span><span class="co">#&gt; dplyr 1.1.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `reframe()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> When switching from `summarise()` to `reframe()`, remember that `reframe()`</span></span>
<span><span class="co">#&gt;   always returns an ungrouped data frame and adjust accordingly.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">maldipickr</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/cpauvert/maldipickr/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; The following spectra are discarded:</span></span>
<span><span class="co">#&gt; species2_E11 species2_E12</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co"># Once checked, the real dereplication can be performed.</span></span>
<span><span class="fu"><a href="../reference/pick_spectra.html">pick_spectra</a></span><span class="op">(</span><span class="va">clusters</span>, discard_regex <span class="op">=</span> <span class="st">"E11"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   name         membership cluster_size   SNR peaks is_reference to_pick</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> species1_G2           1            4  5.09    21 FALSE        FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> species2_E11         <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>           FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> species2_E12         <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>           FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> species3_F7           1            4  4.89    26 FALSE        FALSE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> species3_F8           1            4  5.56    25 TRUE         TRUE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> species3_F9           1            4  5.40    25 FALSE        FALSE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Dumolin C, Aerts M, Verheyde B, Schellaert S, Vandamme T, Van Der
Jeugt F, De Canck E, Cnockaert M, Wieme AD, Cleenwerck I, Peiren J,
Dawyndt P, Vandamme P, &amp; Carlier A. (2019). “Introducing SPeDE:
High-Throughput Dereplication and Accurate Determination of Microbial
Diversity from Matrix-Assisted Laser Desorption–Ionization Time of
Flight Mass Spectrometry Data”. <em>MSystems</em> 4(5). <a href="https://doi.org/10.1128/msystems.00437-19" class="external-link uri">https://doi.org/10.1128/msystems.00437-19</a>.</li>
<li>Strejcek M, Smrhova T, Junkova P &amp; Uhlik O (2018). “Whole-Cell
MALDI-TOF MS versus 16S rRNA Gene Analysis for Identification and
Dereplication of Recurrent Bacterial Isolates.” <em>Frontiers in
Microbiology</em> 9 <a href="https://doi.org/10.3389/fmicb.2018.01294" class="external-link uri">https://doi.org/10.3389/fmicb.2018.01294</a>.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Charlie Pauvert.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
